enclave/core/sgx/once.c

```diff
@@ -5,34 +5,73 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define FUNC_NOT_INVOKED 0
+#define FUNC_BEING_INVOKED 1
+#define FUNC_INVOKED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
-    {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
 
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+    /*
+      Use an atomic-acquire load operation to check if the function has not been
+      invoked. If the function has already been invoked, there is nothing to do.
+      If the function is being invoked, then this thread must wait for the
+      function invocation to complete. Otherwise, this thread can try to take
+      ownership of invoking the function
+    */
+    if (*once != FUNC_INVOKED)
+    {
+        /*
+          Multiple threads could reach here simultaneously after checking
+          whether the function has been invoked or not. Only one of them must
+          now invoke the function and others must wait for the function
+          invocation to complete.
+          To determine who gets to invoke the function, each thread atomically
+          compares the value of once to FUNC_NOT_INVOKED and if equal, sets the
+          value to FUNC_BEING_INVOKED to signal to other threads that the
+          function is being invoked. There return value of
```

Comment:
Change "There return value of" to "The return value"

Reply:
>Are there any better atomic functions to call besides this __sync_val_compare_and_swap ? Here is the reference we used when designing/implementing this: https://gcc.gnu.org/onlinedocs/gcc-7.4.0/gcc/_005f_005fsync-Builtins.html 

This function call seems compatible to the Intel Itanium Processor-specific Application Binary Interface, section 7.4 and not sure if they conform to the Intel IA86/64 ABI. Will investigate some more and update later.

Reply:
@johnkord - Use the modern version i.e. __atomic_compare_exchange_n or  __atomic_compare_exchange. Here is the reference. https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
Trust this helps.